src/main.cpp

```diff
@@ -575,7 +574,178 @@ void affect_nearby_friends_with_alarm(struct Thing *traptng)
+  unsigned long seed = 9377 * game.action_rand_seed + 9439;
+  game.action_rand_seed = LB_RANDOM(4294967296, &seed);
+  if ( !(game.action_rand_seed & 7) && (!*(unsigned short *)&thing->velocity.z.stl ) )
+  {
+    if ( thing_touching_floor(thing) )
+    {
+      long top_cube = get_top_cube_at(thing->mappos.x.stl.num, thing->mappos.y.stl.num, NULL);
+      if ( ((top_cube & 0xFFFFFFFE) != 40) && (top_cube != 39) )
+      {
+        *(unsigned short *)&thing->veloc_push_add.z.stl += 48;
+        thing->state_flags |= TF1_PushAdd;
+      }
+    }
+  }
+  if ( thing_in_wall_at(thing, &pos) )
+  {
+    long blocked_flags = get_thing_blocked_flags_at(thing, &pos);
+    if ( blocked_flags & SlbBloF_WalledX )
+    {
+      angle = thing->move_angle_xy;
+      if ( (angle) && (angle <= 0x400u) )
+        collide = process_boulder_collision(thing, &pos, 1, 0);
+      else
+        collide = process_boulder_collision(thing, &pos, -1, 0);
+    }
+    else if ( blocked_flags & SlbBloF_WalledY )
+    {
+      angle = thing->move_angle_xy;
+      if ( (angle <= 0x200u) || (angle > 0x600u) )
+        collide = process_boulder_collision(thing, &pos, 0, -1);
+      else
+        collide = process_boulder_collision(thing, &pos, 0, 1);
+    }
+    else
+    {
+      collide = 0;
+    }
+    if ( collide != 1 )
+    {
+      if ( (thing->model != ShM_SolidBoulder) && (collide == 0) )
+      {
+        thing->health -= game.boulder_reduce_health_wall;
+      }
+      slide_thing_against_wall_at(thing, &pos, blocked_flags);
+      if ( blocked_flags & SlbBloF_WalledX )
+      {
+        angle = thing->move_angle_xy;
+        if ( (angle) && ( (angle <= 0x200u) || (angle > 0x600u) ) )
+        {
+          unsigned short y = *(unsigned short *)&thing->mappos.y.stl;
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl;
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = y - 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 0 : 0x400;
+        }
+        else
+        {
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl;
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = *(unsigned short *)&thing->mappos.y.stl + 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 0x400 : 0;
+        }
+      }
+      else if ( blocked_flags & SlbBloF_WalledY )
+      {
+        angle = thing->move_angle_xy;
+        if ( (angle) && (angle <= 0x400u) ) 
+        {
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = *(unsigned short *)&thing->mappos.y.stl;
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl + 3 * speed;
```

Comment:
These 3s are subtiles per slab right?

Reply:
Yes. You reckon the 3 is a reference to the number of subtiles per slab and should be STL_PER_SLB?